sklearn/linear_model/coordinate_descent.py

```diff
@@ -696,79 +727,155 @@ def fit(self, X, y, check_input=True):
+                                ensure_2d=False)
+            # Ensure copying happens only once, don't do it again if done above
+            should_copy = self.copy_X and not X_copied
+            if self.precompute is not False:
+                warnings.warn("The solver 'saga' does not use the precomputed "
+                              "Gram matrix", category=UserWarning)
+            X, y, X_offset, y_offset, X_scale, precompute, Xy = \
+                _pre_fit(X, y, None, False, self.normalize,
+                         self.fit_intercept, copy=should_copy,
+                         check_input=check_input)
+            if y.ndim == 1:
+                y = y[:, np.newaxis]
+            if Xy is not None and Xy.ndim == 1:
+                Xy = Xy[:, np.newaxis]
+
+            n_samples, n_features = X.shape
+            n_targets = y.shape[1]
+
+            if not self.warm_start or not hasattr(self, "coef_"):
+                coef_ = np.zeros((n_targets, n_features), dtype=X.dtype,
+                                 order='C')
             else:
-                this_Xy = None
-            _, this_coef, this_dual_gap, this_iter = \
-                self.path(X, y[:, k],
-                          l1_ratio=self.l1_ratio, eps=None,
-                          n_alphas=None, alphas=[self.alpha],
-                          precompute=precompute, Xy=this_Xy,
-                          fit_intercept=False, normalize=False, copy_X=True,
-                          verbose=False, tol=self.tol, positive=self.positive,
-                          X_offset=X_offset, X_scale=X_scale,
-                          return_n_iter=True, coef_init=coef_[k],
-                          max_iter=self.max_iter,
-                          random_state=self.random_state,
-                          selection=self.selection,
-                          check_input=False)
-            coef_[k] = this_coef[:, 0]
-            dual_gaps_[k] = this_dual_gap[0]
-            self.n_iter_.append(this_iter[0])
+                coef_ = self.coef_
+                if coef_.ndim == 1:
+                    coef_ = coef_[np.newaxis, :]
+
+            max_squared_sum = row_norms(X, squared=True).max()
+            self.n_iter_ = []
+
+            for k in range(n_targets):
+                this_coef, this_iter, warm_start_sag = sag_solver(
+                    X, y[:, k], sample_weight=None, loss='squared',
+                    alpha=n_samples * self.alpha * (1 - self.l1_ratio),
+                    beta=n_samples * self.alpha * self.l1_ratio,
+                    max_iter=self.max_iter, tol=self.tol, verbose=0,
+                    random_state=self.random_state, check_input=False,
+                    max_squared_sum=max_squared_sum,
+                    warm_start_mem={'coef': np.expand_dims(coef_[k], axis=1)},
+                    is_saga=True)
+                coef_[k] = this_coef
+                self.n_iter_.append(this_iter)
+                self.dual_gap_ = None
```

Comment:
I think we should compute the `dual_gap_` after the final SAGA iteration (not necessarily use it as a convergence criterion for SAGA itself) but more as a final check to make it possible for the user that the two solvers can find a solution with similar quality.

I don't really know for sure what is the meaning for `tol` in the context of saga. @TomDLT @arthurmensch @agramfort do know if we could provide any kind of guarantee w.r.t. the dual gap when using the natural stopping criterion of the `sag_solver`?

Reply:
From what I see, saga uses the same convergence criteria but without the additional check for the dual gap. Compare https://github.com/scikit-learn/scikit-learn/blob/7bfed173b0f367be8fb9ce31daa4cebcc8cc3a01/sklearn/linear_model/sag_fast.pyx#L501
https://github.com/scikit-learn/scikit-learn/blob/7bfed173b0f367be8fb9ce31daa4cebcc8cc3a01/sklearn/linear_model/cd_fast.pyx#L206-L207